QPS Driver Diff

ncteisen opened the pull request:
Build from #11505.

Running a diff on a few token scenarios for every PR like we do for benchmarks will be very helpful in preventing regressions. Sometimes the microbenchmarks can lie.

Ideally I would like to run this as a separate jenkins job like we already do with trickle and the microbenchmarks.

I will pulled the QPS Driver changes out and have them reviewed so we can test this in the same way we just did for the new trickle diff. At the moment it can't run against master, because the master QPS Driver doesn't support the json_file_out flag. Once the QPS changes are merged I will rename the new script and test

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```

matt-kwong commented:
There's a couple Clang format failures in sanity. Can you ping me when QPS changes necessary to run this on master are in?

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[qps] No significant performance differences
```

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[qps] No significant performance differences
```

ncteisen commented:
QPS changes are merged in and this is running pretty well. Ready for review

matt-kwong commented on test/cpp/qps/qps_diff.py:
```diff
@@ -0,0 +1,164 @@
+# limitations under the License.
+""" Computes the diff between two qps runs and outputs significant results """
+
+import shutil
```
Alphabetize the imports

ncteisen replied:
done

matt-kwong commented on test/cpp/qps/qps_diff.py:
```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python2.7
```
Should this file be in `test/cpp`? It looks like it might fit in better in `tools/profiling`.

ncteisen replied:
done.

matt-kwong commented on test/cpp/qps/qps_diff.py:
```diff
@@ -0,0 +1,164 @@
+    os.path.dirname(sys.argv[0]), '..', '..', '..', 'tools', 'run_tests', 'python_utils'))
+import comment_on_pr
+
+_SCENARIOS = {
```
It might be worth it to format this or keep this in a separate JSON file if this list is expected to grow.

ncteisen replied:
Good idea, done

matt-kwong commented on test/cpp/qps/qps_diff.py:
```diff
@@ -0,0 +1,164 @@
+  'large-message-throughput': '{"scenarios":[{"name":"large-message-throughput", "spawn_local_worker_count": -2, "warmup_seconds": 30, "benchmark_seconds": 270, "num_servers": 1, "server_config": {"async_server_threads": 1, "security_params": null, "server_type": "ASYNC_SERVER"}, "num_clients": 1, "client_config": {"client_type": "ASYNC_CLIENT", "security_params": null, "payload_config": {"simple ...
+  'multi-channel-64-KiB': '{"scenarios":[{"name":"multi-channel-64-KiB", "spawn_local_worker_count": -3, "warmup_seconds": 30, "benchmark_seconds": 270, "num_servers": 1, "server_config": {"async_server_threads": 31, "security_params": null, "server_type": "ASYNC_SERVER"}, "num_clients": 2, "client_config": {"client_type": "ASYNC_CLIENT", "security_params": null, "payload_config": {"simple_params ...
+}
+
```
2 newlines in between each function defined

ncteisen replied:
done

matt-kwong commented on test/cpp/qps/qps_diff.py:
```diff
@@ -0,0 +1,164 @@
+    '--loops',
+    type=int,
+    default=4,
+    help='Number of times to loops the benchmarks. More loops cuts down on noise'
```
"Number of loops for each benchmark" sounds better

ncteisen replied:
done

matt-kwong commented on test/cpp/qps/qps_diff.py:
```diff
@@ -0,0 +1,164 @@
+    return None
+
+def main(args):
+
```
Get rid of newline here

ncteisen replied:
done

matt-kwong commented on test/cpp/qps/qps_diff.py:
```diff
@@ -0,0 +1,164 @@
+  return args
+
+def _make_cmd(jobs):
+  return ['make', '-j', '%d' % jobs, 'qps_json_driver', 'qps_worker',]
```
remove last `,`

ncteisen replied:
done

matt-kwong commented on test/cpp/qps/qps_diff.py:
```diff
@@ -0,0 +1,164 @@
+
+if __name__ == '__main__':
+  args = _args()
+  main(args);
```
Get rid of `;`

ncteisen replied:
too much c in my brain... done

ncteisen commented:
Addressed all comments.

With respect to the output from driver.cc; I am passing in `spawn_local_worker_count` to the benchmark. This will create the workers before starting the runs. The warning about QPS_WORKERS can be ignored

grpc-jenkins commented:
```
[qps] No significant performance differences
```

grpc-jenkins commented:
```
[qps] No significant performance differences
```

grpc-jenkins commented:
```
[qps] No significant performance differences
```

matt-kwong commented on tools/profiling/qps/qps_scenarios.py:
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env python2.7
```
Don't need this, since this file will never be called directly.

matt-kwong commented on tools/profiling/qps/qps_diff.py:
```diff
@@ -0,0 +1,169 @@
+  for sn in scenarios:
+    old_data[sn] = []
+    new_data[sn] = []
+    for i in range(0, loops):
```
This can just be `range(loops):`

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] No significant performance differences
```
